sdks/go/pkg/beam/core/runtime/exec/sdf_invokers.go

```diff
@@ -0,0 +1,302 @@
+	case reflectx.Func1x1:
+		n.call = func(elms *FullValue) interface{} {
+			return fnT.Call1x1(elms.Elm)
+		}
+	case reflectx.Func2x1:
+		n.call = func(elms *FullValue) interface{} {
+			return fnT.Call2x1(elms.Elm, elms.Elm2)
+		}
+	default:
+		switch len(n.fn.Param) {
+		case 1:
+			n.call = func(elms *FullValue) interface{} {
+				n.args[0] = elms.Elm
+				return n.fn.Fn.Call(n.args)[0]
+			}
+		case 2:
+			n.call = func(elms *FullValue) interface{} {
+				n.args[0] = elms.Elm
+				n.args[1] = elms.Elm2
+				return n.fn.Fn.Call(n.args)[0]
+			}
+		default:
+			return errors.Errorf("CreateInitialRestriction fn %v has unexpected number of parameters: %v",
+				n.fn.Fn.Name(), len(n.fn.Param))
+		}
+	}
+
+	return nil
+}
+
+// Invoke calls CreateInitialRestriction with the given FullValue as the element
+// and returns the resulting restriction.
+func (n *cirInvoker) Invoke(elms *FullValue) (rest interface{}) {
+	return n.call(elms)
+}
+
+// Reset zeroes argument entries in the cached slice to allow values to be
+// garbage collected after the bundle ends.
+func (n *cirInvoker) Reset() {
+	for i := range n.args {
+		n.args[i] = nil
+	}
+}
+
+// srInvoker is an invoker for SplitRestriction.
+type srInvoker struct {
+	fn   *funcx.Fn
+	args []interface{} // Cache to avoid allocating new slices per-element.
+	call func(elms *FullValue, rest interface{}) (split interface{})
+}
+
+func newSplitRestrictionInvoker(fn *funcx.Fn) (*srInvoker, error) {
+	n := &srInvoker{
+		fn:   fn,
+		args: make([]interface{}, len(fn.Param)),
+	}
+	if err := n.initCallFn(); err != nil {
+		return nil, errors.WithContext(err, "sdf SplitRestriction invoker")
+	}
+	return n, nil
+}
+
+func (n *srInvoker) initCallFn() error {
+	// Expects a signature of the form:
+	// (key?, value, restriction) []restriction
+	// TODO(BEAM-9643): Link to full documentation.
+	switch fnT := n.fn.Fn.(type) {
+	case reflectx.Func2x1:
+		n.call = func(elms *FullValue, rest interface{}) interface{} {
+			return fnT.Call2x1(elms.Elm, rest)
+		}
+	case reflectx.Func3x1:
+		n.call = func(elms *FullValue, rest interface{}) interface{} {
+			return fnT.Call3x1(elms.Elm, elms.Elm2, rest)
+		}
+	default:
+		switch len(n.fn.Param) {
+		case 2:
+			n.call = func(elms *FullValue, rest interface{}) interface{} {
+				n.args[0] = elms.Elm
+				n.args[1] = rest
+				return n.fn.Fn.Call(n.args)[0]
+			}
+		case 3:
+			n.call = func(elms *FullValue, rest interface{}) interface{} {
+				n.args[0] = elms.Elm
+				n.args[1] = elms.Elm2
+				n.args[2] = rest
+				return n.fn.Fn.Call(n.args)[0]
+			}
+		default:
+			return errors.Errorf("SplitRestriction fn %v has unexpected number of parameters: %v",
+				n.fn.Fn.Name(), len(n.fn.Param))
+		}
+	}
+	return nil
+}
+
+// Invoke calls SplitRestriction given a FullValue containing an element and
+// the associated restriction, and returns a slice of split restrictions.
+func (n *srInvoker) Invoke(elms *FullValue, rest interface{}) (split []interface{}) {
```

Comment:
eg. This is a great use of a named return value to add clarity. 

I'd call it "splits" though, since it's a slice and there could be more than one.

Reply:
Agreed, "splits" makes more sense.